mqtt.go

```diff
@@ -175,442 +235,305 @@ func (mc *mqttConn) subscribe(subs []mqttSubscription) error {
+	} else {
+		if conn, err = mqttDialer.Dial("tcp", addr); err == nil {
+		} else {
+			logError("MQTT dialer failed: %s", err.Error())
 			return nil
 		}
-
-		msg := fmt.Sprintf("writeBulkData() received packet that does not match the Ack packet ID (%d) and the expected packet ID (%d).",
-			ack.PacketID, p.PacketID)
-		logInfo("[MQTT] %s", msg)
-		mc.reportError(fmt.Errorf("synced bulk data delivery unmatched Ack Packet: %s", msg))
 	}
 
-	return errors.New("did not receive ack packet until timeout and bulk data dropped")
+	return &mqttConn{
+		conn:   conn,
+		stream: packet.NewStream(conn, conn),
+		// This map will grow to the size of the number of requests that are
+		// unanswered
+		packIDToChan: make(map[uint16]responseChan),
+		connRespCh:   make(responseChan),
+		pingRespCh:   make(responseChan),
+		pubHandleCh:  make(responseChan),
+	}
 }
 
-func (mc *mqttConn) sendKeyValueUpdate(kvSync *keyValueSync) error {
-	// Always QoS1
-	qos := packet.QOSAtLeastOnce
+// Called by the client to send packets to the server.
+func (conn *mqttConn) SendPacket(ctx context.Context,
+	p packet.Packet) (responseChan, error) {
 
-	payload, _ := json.Marshal(kvSync)
-
-	p := packet.NewPublishPacket()
-	p.PacketID = mc.getPacketID()
-	p.Message = packet.Message{
-		Topic:   fmt.Sprintf("/devices/%d/kv", mc.dc.DeviceID),
-		QOS:     qos,
-		Payload: payload,
+	ch := conn.getResponseChan(p)
+	//logInfo("[MQTT]: sending packet: ", p)
+	if err := conn.stream.Write(p); err != nil {
+		logError("[MQTT] failed to send %s packet: %s", p.Type(), err.Error())
+		return nil, err
 	}
 
-	for count := uint(1); count <= maxKeyValueUpdateAttempts; count++ {
-		mc.outPacket <- p
-
-		logInfo("[MQTT] key-value update attempt #%d for packet ID %d", count, p.PacketID)
-		logInfo("[MQTT] waiting for PUBACK for packet ID %d", p.PacketID)
-
-		select {
-		case <-time.After(keyValueUpdateRetryInterval):
-			logError("[MQTT] did not receive PUBACK for packet ID %d within %v", p.PacketID, keyValueUpdateRetryInterval)
-
-		case ack := <-mc.puback:
-			if ack.PacketID == p.PacketID {
-				logInfo("[MQTT] received PUBACK for packet ID %d", ack.PacketID)
-				return nil
-			}
-			// TBD: Something is really wrong if packet ID does not match. What to do?
-		}
-
-		p.Dup = true
+	if err := conn.stream.Flush(); err != nil {
+		logError("[MQTT] failed to flush %s packet: %s", p.Type(), err.Error())
+		return nil, err
 	}
 
-	return errors.New("key-value update dropped")
+	return ch, nil
 }
 
-func (dc *DeviceContext) openMQTTConn(cmdQueue chan<- *DeviceCommand, evtQueue <-chan *DeviceEvent, evtBulkDataQueue <-chan *DeviceBulkData, evtSyncedBulkDataCh <-chan *DeviceSyncedBulkData, kvSyncQueue chan<- *keyValueSync, kvPushQueue <-chan *keyValueSync) (*mqttConn, error) {
-	mc := &mqttConn{
-		dc:   dc,
-		subs: make(map[string]mqttSubscription),
-	}
+// Called by the client to start listening for packets from the server
+func (conn *mqttConn) RunPacketReader(wg *sync.WaitGroup) {
+	wg.Add(1)
```

Comment:
This method can call more than once?

Reply:
Ah. No, only once. It shouldn't be exported.